File: SPD-classes/src/main/java/com/watabou/utils/Point.java
/*
 * Pixel Dungeon
 * Copyright (C) 2012-2015 Oleg Dolya
 *
 * Shattered Pixel Dungeon
 * Copyright (C) 2014-2024 Evan Debenham
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>
 */

package com.watabou.utils;

public class Point {

	public int x;
	public int y;
	
	public Point() {
	}
	
	public Point( int x, int y ) {
		this.x = x;
		this.y = y;
	}
	
	public Point( Point p ) {
		this.x = p.x;
		this.y = p.y;
	}
	
	public Point set( int x, int y ) {
		this.x = x;
		this.y = y;
		return this;
	}
	
	public Point set( Point p ) {
		x = p.x;
		y = p.y;
		return this;
	}
	
	public Point clone() {
		return new Point( this );
	}
	
	public Point scale( float f ) {
		this.x *= f;
		this.y *= f;
		return this;
	}
	
	public Point offset( int dx, int dy ) {
		x += dx;
		y += dy;
		return this;
	}
	
	public Point offset( Point d ) {
		x += d.x;
		y += d.y;
		return this;
	}

	public boolean isZero(){
		return x == 0 && y == 0;
	}

	public float length() {
		return (float)Math.sqrt( x * x + y * y );
	}

	public static float distance( Point a, Point b ) {
		float dx = a.x - b.x;
		float dy = a.y - b.y;
		return (float)Math.sqrt( dx * dx + dy * dy );
	}
	
	@Override
	public boolean equals( Object obj ) {
		if (obj instanceof Point) {
			Point p = (Point)obj;
			return p.x == x && p.y == y;
		} else {
			return false;
		}
	}
}

File: pixel-dungeon-web/src/utils/Point.js
/*
 * Pixel Dungeon Web Port
 * Ported from Shattered Pixel Dungeon
 * Copyright (C) 2012-2015 Oleg Dolya
 * Copyright (C) 2014-2025 Evan Debenham
 *
 * 2D Integer Point
 * Source: SPD-classes/src/main/java/com/watabou/utils/Point.java
 */

export class Point {
    constructor(x = 0, y = 0) {
        if (typeof x === 'object' && x instanceof Point) {
            // Copy constructor: Point.java:37-40
            this.x = x.x;
            this.y = x.y;
        } else {
            // Normal constructor: Point.java:32-35
            this.x = x;
            this.y = y;
        }
    }

    /**
     * Set coordinates
     * Source: Point.java:42-52
     */
    set(x, y) {
        if (typeof x === 'object' && x instanceof Point) {
            this.x = x.x;
            this.y = x.y;
        } else {
            this.x = x;
            this.y = y;
        }
        return this;
    }

    /**
     * Create a copy of this point
     * Source: Point.java:54-56
     */
    clone() {
        return new Point(this.x, this.y);
    }

    /**
     * Scale by a factor
     * Source: Point.java:58-62
     */
    scale(f) {
        this.x = Math.floor(this.x * f);
        this.y = Math.floor(this.y * f);
        return this;
    }

    /**
     * Offset by delta
     * Source: Point.java:64-74
     */
    offset(dx, dy) {
        if (typeof dx === 'object' && dx instanceof Point) {
            this.x += dx.x;
            this.y += dx.y;
        } else {
            this.x += dx;
            this.y += dy;
        }
        return this;
    }

    /**
     * Check if point is at origin
     * Source: Point.java:76-78
     */
    isZero() {
        return this.x === 0 && this.y === 0;
    }

    /**
     * Calculate length from origin
     * Source: Point.java:80-82
     */
    length() {
        return Math.sqrt(this.x * this.x + this.y * this.y);
    }

    /**
     * Calculate distance between two points
     * Source: Point.java:84-88
     */
    static distance(a, b) {
        const dx = a.x - b.x;
        const dy = a.y - b.y;
        return Math.sqrt(dx * dx + dy * dy);
    }

    /**
     * Equality check
     * Source: Point.java:90-98
     */
    equals(obj) {
        if (obj instanceof Point) {
            return obj.x === this.x && obj.y === this.y;
        }
        return false;
    }

    /**
     * String representation
     */
    toString() {
        return `Point(${this.x}, ${this.y})`;
    }
}
